src/graphics/text.c

```diff
@@ -328,48 +330,98 @@ static int number_to_string(uint8_t *str, int value, char prefix, const char *po
 {
-    uint8_t str[NUMBER_BUFFER_LENGTH];
-    number_to_string(str, value, prefix, postfix, 0);
-    return text_draw(str, x_offset, y_offset, font, 0);
+    int width;
+    if (c == ' ' || c == '_' || letter_id < 0) {
+        width = def->space_width;
+    } else {
+        const image *img = image_letter(letter_id);
+        int height = def->image_y_offset(c, img->height, def->line_height);
+        image_draw_letter(def->font, letter_id, current_x, y - height, color);
+        width = def->letter_spacing + img->width;
+    }
+    if (input_cursor.capture && input_cursor.position == input_cursor.cursor_position) {
+        if (!input_cursor.seen) {
+            input_cursor.width = width;
+            input_cursor.x_offset = current_x - x;
+            input_cursor.seen = 1;
+        }
+    }
+    return width;
 }
 
-int text_draw_number_with_separator(int value, char prefix, const char* postfix, int x_offset, int y_offset, font_t font)
+int text_draw_number(int value, char prefix, const char *postfix, int x, int y, font_t font, color_t color)
 {
-    uint8_t str[NUMBER_BUFFER_LENGTH];
-    number_to_string(str, value, prefix, postfix, locale_number_thousands_separator());
-    return text_draw(str, x_offset, y_offset, font, 0);
-}
+    const font_definition *def = font_definition_for(font);
+    int current_x = x;
 
-int text_draw_number_colored(
-    int value, char prefix, const char *postfix, int x_offset, int y_offset, font_t font, color_t color)
-{
-    uint8_t str[NUMBER_BUFFER_LENGTH];
-    number_to_string(str, value, prefix, postfix, 0);
-    return text_draw(str, x_offset, y_offset, font, color);
-}
+    if (prefix > ' ') {
+        int num_bytes = 1;
+        int letter_id = font_letter_id(def, &prefix, &num_bytes);
+        int width = put_one_char(def, prefix, letter_id, x, current_x, y, color);
+        current_x += width;
+        input_cursor.position += num_bytes;
+    }
 
-int text_draw_number_colored_with_separator(
-    int value, char prefix, const char* postfix, int x_offset, int y_offset, font_t font, color_t color)
-{
-    uint8_t str[NUMBER_BUFFER_LENGTH];
-    number_to_string(str, value, prefix, postfix, locale_number_thousands_separator());
-    return text_draw(str, x_offset, y_offset, font, color);
+    uint8_t buffer[NUMBER_BUFFER_LENGTH];
+    int length = string_from_int(buffer, value, 0);
+    const char *str = buffer;
+    if (input_cursor.capture) {
+        str += input_cursor.text_offset_start;
+        length = input_cursor.text_offset_end - input_cursor.text_offset_start;
+    }
+
+    int separator_pixels = config_get(CONFIG_UI_DIGIT_SEPARATOR_PIXELS);
+    
+    while (length > 0) {
+        int num_bytes = 1;
+
+        if (*str >= ' ') {
+            int letter_id = font_letter_id(def, str, &num_bytes);
+            int width = put_one_char(def, *str, letter_id, x, current_x, y, color);
+            current_x += width + ((length == 4 || length == 7) ? separator_pixels : 0);
```

Comment:
Careful: relying on fixed length will in some cases yield the wrong result when dealing with negative numbers.

Reply:
I'm having trouble thinking of a situation where this calculation won't work. Because we're writing characters starting at the left and counting down length, length means how many characters remain on the right. We would always hit a separator when we have 3 characters to the right and 6 characters to the right (+1 because we're writing the separator at the bottom of the loop instead of the top). If you can point to a counter-example where this logic doesn't hold I can redo this algorithm.

Reply:
You're right, I was thinking about this wrong.